### PR TITLE
Major overhaul of Lazy[_], Foldable[_], etc.

### PR DESCRIPTION
**Pull request by non**
This commit does a number of significant things:

 * Removes the `Fold[_]` implementation entirely.
 * Renames `Lazy[_]` to `Eval[_]`
 * Support monadic recursion over `Eval[_]` via trampoline
 * Remove partialFold &co
 * Update foldRight definition to be more standard
 * Fix documentation of `Lazy[_]`, `Foldable[_]`, etc.
 * Fix everything else to work with these changes

The trampoline used in `Eval[_]` is specialized and should be more efficient than
the standard encoding (although benchmarks are still needed to confirm this).
The semantics behind `Eval[_]` are hopefully a bit clearer, and the documentation
tries to make it clear exactly what one can expect.

I think this is a huge improvement over the status quo. The whole undertaking
was inspired by Miles' desire to have easier derivations for Foldable, so I
hope that this takes care of that. It also tries to address many of the
concerns around `Lazy[_]` that were brought up (many of which were ultimately
correct).

Review by @tpolecat, @milessabin, and anyone else who is interested.

### CODE REVIEW

**Comment by milessabin** [comment]
Huge :+1: from me ... this is a big improvement :-)

**Comment by milessabin** [comment]
My only comment is that it might be an idea to rename `Lazy` to `ByNeed` for consistency with `ByName` and also with the companion constructor `byNeed`. It would also remove a name conflict with shapeless's `Lazy`.

**Comment by tpolecat** [comment]
Agree with the naming thing Miles mentioned, but otherwise looks great. :+1:

File: core/src/main/scala/cats/Eval.scala [hunk 1]
```diff
@@ -0,0 +1,275 @@
+package cats
+
+import scala.annotation.tailrec
+import cats.syntax.all._
+
+/**
+ * Eval is a monad which controls evaluation.
+ *
+ * This types wraps a value (or a computation that produces a value)
```
Comment (rintcius) [inline core/src/main/scala/cats/Eval.scala]: types -> type

File: laws/src/main/scala/cats/laws/discipline/Arbitrary.scala [hunk 2]
```diff
@@ -38,9 +38,9 @@ object arbitrary {
   implicit def optionTArbitrary[F[_], A](implicit F: ArbitraryK[F], A: Arbitrary[A]): Arbitrary[OptionT[F, A]] =
     Arbitrary(F.synthesize[Option[A]].arbitrary.map(OptionT.apply))
 
-  implicit def foldArbitrary[A](implicit A: Arbitrary[A]): Arbitrary[Fold[A]] =
-    Arbitrary(Gen.oneOf(getArbitrary[A].map(Fold.Return(_)), getArbitrary[A => A].map(Fold.Continue(_)), Gen.const(Fold.Pass[A])))
-
-  implicit def lazyArbitrary[A](implicit A: Arbitrary[A]): Arbitrary[Lazy[A]] =
-    Arbitrary(Gen.oneOf(A.arbitrary.map(Lazy.eager), A.arbitrary.map(a => Lazy.byName(a)), A.arbitrary.map(a => Lazy.byNeed(a))))
+  implicit def lazyArbitrary[A](implicit A: Arbitrary[A]): Arbitrary[Eval[A]] =
```
Comment (rintcius) [inline laws/src/main/scala/cats/laws/discipline/Arbitrary.scala]: rename to `evalArbitrary` now?

File: laws/src/main/scala/cats/laws/discipline/ArbitraryK.scala [hunk 3]
```diff
@@ -45,10 +45,10 @@ object ArbitraryK {
   implicit val list: ArbitraryK[List] =
     new ArbitraryK[List] { def synthesize[A: Arbitrary]: Arbitrary[List[A]] = implicitly }
 
-  implicit val lazy_ : ArbitraryK[Lazy] =
-    new ArbitraryK[Lazy] {
-      def synthesize[A](implicit A: Arbitrary[A]): Arbitrary[Lazy[A]] =
-        Arbitrary(A.arbitrary.map(Lazy(_)))
+  implicit val lazy_ : ArbitraryK[Eval] =
```
Comment (rintcius) [inline laws/src/main/scala/cats/laws/discipline/ArbitraryK.scala]: rename `lazy_`?

File: laws/src/main/scala/cats/laws/discipline/EqK.scala [hunk 4]
```diff
@@ -34,8 +34,8 @@ object EqK {
   implicit val list: EqK[List] =
     new EqK[List] { def synthesize[A: Eq]: Eq[List[A]] = implicitly }
 
-  implicit val lazy_ : EqK[Lazy] =
-    new EqK[Lazy] { def synthesize[A: Eq]: Eq[Lazy[A]] = implicitly }
+  implicit val lazy_ : EqK[Eval] =
```
Comment (rintcius) [inline laws/src/main/scala/cats/laws/discipline/EqK.scala]: also rename?

File: tests/src/test/scala/cats/tests/FoldableTests.scala [hunk 5]
```diff
@@ -1,18 +1,61 @@
...
+      fa.find(_ > n)   shouldBe iterator(fa).find(_ > n)
+      fa.exists(_ > n) shouldBe iterator(fa).exists(_ > n)
+      fa.forall(_ > n) shouldBe iterator(fa).forall(_ > n)
+      fa.filter_(_ > n) shouldBe iterator(fa).filter(_ > n).toList
+      fa.dropWhile_(_ > n) shouldBe iterator(fa).dropWhile(_ > n).toList
+    }
+  }
+
+  test("toList/isEmpty/nonEmpty") {
+    forAll { (fa: F[Int], n: Int) =>
```
Comment (rintcius) [inline tests/src/test/scala/cats/tests/FoldableTests.scala]: `n` not used, it seems

File: tests/src/test/scala/cats/tests/FoldableTests.scala [hunk 6]
```diff
@@ -1,18 +1,61 @@
...
+      fa.toList shouldBe iterator(fa).toList
+      fa.isEmpty shouldBe iterator(fa).isEmpty
+      fa.nonEmpty shouldBe iterator(fa).nonEmpty
+    }
+  }
+
+  test("exists") {
+    forAll { (fa: F[Int], n: Int) =>
```
Comment (rintcius) [inline tests/src/test/scala/cats/tests/FoldableTests.scala]: TODO

File: tests/src/test/scala/cats/tests/LazyTests.scala [hunk 7]
```diff
@@ -5,7 +5,7 @@ import scala.math.min
 
 // TODO: monad laws
 
-class LazyTests extends CatsSuite {
+class EvalTests extends CatsSuite {
```
Comment (rintcius) [inline tests/src/test/scala/cats/tests/LazyTests.scala]: file could be renamed too

**Comment by non** [comment]
@rintcius Thanks for your thorough review. I'll go through and fix those.

@milessabin, @tpolecat I'm going to try out `Now`, `Later`, and `Always`, which I think read a bit better.

**Comment by milessabin** [comment]
Big :+1: on this :-)

**Comment by rintcius** [comment]
@non :+1:

**Comment by adelbertc** [comment]
This is pretty awesome :-) :+1:

**Pull request merged**